### ISSUE TITLE
[Modal; Vanilla] Carry React focus trap functionality to Vanilla implementation

### ISSUE THREAD

**Issue by elizabethsjudd** [carbon-design-system/carbon#2939]
Watson Health is using the Vanilla JavaScript of Carbon.  The React version of Modal traps focus within the Modal but the vanilla does not appear to have this functionality.  Due to compliance regulations we need this functionality carried over.

## Environment

> Operating system

Mac

> Browser

Chrome

> Automated testing tool and ruleset

> Assistive technology used to verify

Keyboard

## Detailed description

> What version of the Carbon Design System are you using?

v10.2.0

> What did you expect to happen?

When I'm tabbing within the Modal, focus should remain within the modal and not be able to move to other elements within the page

> What happened instead?

I could tab to elements on the page when the modal is open

> What WCAG 2.1 checkpoint does the issue violate?

https://www.w3.org/TR/wai-aria-practices-1.1/#dialog_modal

## Steps to reproduce the issue

1. Open a modal
2. tab until you've gone through all of the elements in the modal
3. press tab
4. element outside of modal has focus

> Please create a reduced test case in CodeSandbox
>
> - Style and vanilla JS:
>   https://codesandbox.io/s/modal-2vjuw

**Comment by dakahn** [comment]
For the React modal component we used focus-trap-react
https://github.com/davidtheclark/focus-trap-react

There's also a vanilla version! 
https://github.com/davidtheclark/focus-trap

**Comment by elizabethsjudd** [comment]
@dakahn I can take a stab at this if you want to assign it to me

**Comment by elizabethsjudd** [comment]
@asudoh Looks like there is something in the JS code that is already trying to trap the focus but 1. doesn't seem to be consistent (i couldn't get it to work in WH) and 2. it looks like it's focusing a "hidden" element.  Where should i be looking for this code? Are we planning on removing that in favor of `trap-focus`?

**Comment by asudoh** [comment]
@elizabethsjudd The first element to put focus on in modal tends to depend on the context of application. Given that, we let application set `data-modal-primary-focus` attribute on such element. The ultimate implementation of this feature for vanilla can be seen at [here](https://github.com/carbon-design-system/carbon-custom-elements/pull/22/files#diff-7b81c66d6f3e9c4b82c40ee5ecd30e50).

**Comment by elizabethsjudd** [comment]
@asudoh I have it working locally by using the `data-modal-primary-focus` attribute to allow the developers to chose where it starts it's focus.  Are you saying you'd prefer me to use that you have in the `custom-elements` repo instead of the plugin that @dakahn mentioned to use?

**Comment by asudoh** [comment]
@elizabethsjudd Thank you for double-checking! Basically yes instead of the 3rd party library (there have been a few reports that they have had to disable that feature due to Carbon integration issues). BTW one thing that vanilla has but my link doesn't (yet) is "floating menu" check: https://github.com/carbon-design-system/carbon/blob/v10.3.0/packages/components/src/components/modal/modal.js#L181-L183

**Comment by elizabethsjudd** [comment]
@carbon-design-system/developers if you assign this to me. I can take a look at it in the next couple weeks
